codalab/rest/users.py

```diff
@@ -1,10 +1,133 @@
 """
-Worksheets REST API Users Views
+Worksheets REST API Users Views.
+
+TODO(sckoo): Factor out the reusable parts of the JSON API views.
 """
-from bottle import get
+from bottle import get, route, response, request, local, HTTPResponse
+from marshmallow import ValidationError
+from marshmallow_jsonapi import Schema, fields
+
+from codalab.common import UsageError
+from codalab.lib import formatting
+from codalab.server.authenticated_plugin import AuthenticatedPlugin
+from codalab.server.json_api_plugin import JsonApiPlugin
+
+
+PUBLIC_USER_FIELDS = ('id', 'user_name', 'email', 'first_name', 'last_name',
+                      'date_joined', 'affiliation', 'url')
+
+
+class DataSize(fields.Field):
+    def _serialize(self, value, attr, obj):
+        return formatting.size_str(value)
+
+    def _deserialize(self, value, attr, data):
+        return formatting.parse_size(value)
+
+
+class Duration(fields.Field):
+    def _serialize(self, value, attr, obj):
+        return formatting.duration_str(value)
+
+    def _deserialize(self, value, attr, data):
+        return formatting.parse_duration(value)
+
+
+class UserSchema(Schema):
+    id = fields.String(dump_only=True, attribute='user_id')
+    user_name = fields.String()
+    email = fields.Email(dump_only=True)
+    first_name = fields.String(allow_none=True)
+    last_name = fields.String(allow_none=True)
+    affiliation = fields.String(allow_none=True)
+    url = fields.Url(allow_none=True)
+    time_quota = Duration(dump_only=True)
+    time_used = Duration(dump_only=True)
+    disk_quota = DataSize(dump_only=True)
+    disk_used = DataSize(dump_only=True)
+    last_login = fields.LocalDateTime("%c", dump_only=True)
+    date_joined = fields.LocalDateTime("%c", dump_only=True)
+
+    class Meta:
+        type_ = 'users'
+
+
+@get('/user', apply=[JsonApiPlugin(), AuthenticatedPlugin()])
+def fetch_authenticated_user():
+    """Fetch authenticated user."""
+    fieldset = request.jsonapi.fields.get('users', None)
+    return UserSchema(only=fieldset).dump(request.user).data
+
 
+@route('/user', method=['PUT', 'PATCH'], apply=[JsonApiPlugin(), AuthenticatedPlugin()])
+def update_authenticated_user():
+    """Update one or multiple fields of the authenticated user."""
+    # Load update request data
+    try:
+        user_info, errors = UserSchema(strict=True).load(request.json, partial=True)
+    except ValidationError as e:
+        response.status = '403 Forbidden'
+        return e.messages
 
-# TODO(skoo): user creation endpoint first
-@get('/users')
+    # Patch in user_id manually (do not allow requests to change id)
+    user_info['user_id'] = request.user.user_id
+
+    # Ensure that user name is not taken
+    if (user_info.get('user_name', request.user.user_name) != request.user.user_name and
+        local.model.user_exists(user_info['user_name'], None)):
+        response.status = '403 Forbidden'
```

Comment:
You have wildly different error handling code. In one place you just set status, in another you raise a UsageError, in a third (below) you return an HTTPResponse with code 404. Please be consistent. Personally, I would do away with the JSON API errors return value, and simply use HTTP Error codes and body. See these two pull requests:

https://github.com/codalab/codalab-cli/pull/394
https://github.com/codalab/codalab-worksheets/pull/120